File: StructuralPatterns/Decorator/Adventurer.py
from abc import ABC, abstractmethod

# 冒險者界面(component) - 規範冒險者的功能
class Adventurer(ABC):
    @abstractmethod
    def attack(self):
        pass

class Archer(Adventurer):
    def __init__(self, name):
        self.name = name

    def attack(self):
        print(f"Archer attack by {self.name}")

# 稱號界面(decorator)
class Title(Adventurer, ABC):
    def __init__(self, adventurer):
        self.adventurer = adventurer

    def attack(self):
        self.adventurer.attack()

class TitleStrong(Title):
    def __init__(self, adventurer):
        super().__init__(adventurer)

    def attack(self):
        print("Powerful", end=" ")
        super().attack()

class TitleAgile(Title):
    def __init__(self, adventurer):
        super().__init__(adventurer)

    def attack(self):
        print("Fast", end=" ")
        super().attack()

    def useFlash(self):
        print(" use Fast skill")

class TitleFire(Title):
    def __init__(self, adventurer):
        super().__init__(adventurer)

    def attack(self):
        print("Burning", end=" ")
        super().attack()

    def useFire(self):
        print(" use Burning skill")

File: DESIGN-CSharp/Decorator/Adventurer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Decorator
{
    //冒險者界面(component) - 規範冒險者的功能
    public abstract class Adventurer
    {
        public abstract void attack();
    }

    class Archer : Adventurer
    {
        public string name { get; set; }

        public Archer(string name) 
        {
            this.name = name;
        }

        public override void attack()
        {
            Console.WriteLine("Archar attack by " + this.name);
        }
    }

    //稱號界面(decorator)
    public abstract class Title : Adventurer 
    {
        protected Adventurer adventurer; //被裝飾的冒險者

        public Title(Adventurer adventurer) 
        {
            this.adventurer = adventurer;
        }

        public override void attack()
        {
            this.adventurer.attack();
        }
    }

    public class TitleStrong: Title
    {

        public TitleStrong(Adventurer adventurer) : base(adventurer)
        {
        }

        public override void attack()
        {
            Console.WriteLine(" Powerful ");
            base.attack();
        }
    }

    public class TitleAgile : Title
    {

        public TitleAgile(Adventurer adventurer) : base(adventurer)
        {
        }

        public override void attack()
        {
            Console.WriteLine(" Fast ");
            base.attack();
        }

        //該稱號的專屬技能
        public void useFlash() 
        {
            Console.WriteLine(" use Fast skill ");
        }
    }

    public class TitleFire : Title
    {

        public TitleFire(Adventurer adventurer) : base(adventurer)
        {
        }

        public override void attack()
        {
            Console.WriteLine(" Burning ");
            base.attack();
        }

        //該稱號的專屬技能
        public void useFire()
        {
            Console.WriteLine(" use Burning skill ");
        }
    }
}
